src/Tracepoints.jl

```diff
@@ -0,0 +1,497 @@
+all tracepoints in `mod` are disabled. If `mod` is unspecified, then all
+tracepoints in all loaded modules are disabled.
+
+This function is the counterpart to [`enable!`](ref), which can be called to
+undo a call to `disable!`.
+"""
+function disable! end
+
+function set!(f::Union{Base.Callable, Ptr{Cvoid}}; kwargs...)
+    for mod in values(Base.loaded_modules)
+        if isdefined(mod, :__tracepoints_lock__)
+            set!(f, mod; kwargs...)
+        end
+    end
+end
+function set!(f::Union{Base.Callable, Ptr{Cvoid}}, mod::Module; kwargs...)
+    categories = lock(mod.__tracepoints_lock__) do
+        collect(keys(mod.__tracepoints_specs__))
+    end
+    for category in categories
+        set!(f, mod, category; kwargs...)
+    end
+end
+function set!(payload::Ptr{Cvoid}, mod::Module, category::Symbol; abi::Symbol, incompatible::Symbol=:warn)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        for spec in values(tracepoints[category])
+            validate_abi(spec, abi, incompatible) || continue
+            spec.payload[] = payload
+        end
+    end
+end
+function set!(payload::Ptr{Cvoid}, mod::Module, category::Symbol, kind::Symbol; abi::Symbol, incompatible::Symbol=:warn)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    lock(mod_lock) do
+        spec = tracepoints[category][kind]
+        validate_abi(spec, abi, incompatible) || return
+        spec.payload[] = payload
+    end
+end
+function set!(f, mod::Module, category::Symbol; abi::Symbol=:all, incompatible::Symbol=:warn)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    funcs = mod.__tracepoints_funcs__
+    specs = lock(mod_lock) do
+        tracepoints[category]
+    end
+    for (kind, spec) in specs
+        validate_abi(spec, abi, incompatible) || continue
+        ptr = generate_probe_fptr(f, spec)
+        lock(mod_lock) do
+            spec.payload[] = ptr
+            funcs[category][kind] = f
+        end
+    end
+end
+function set!(f, mod::Module, category::Symbol, kind::Symbol; abi::Symbol=:all, incompatible::Symbol=:warn)
+    mod_lock = mod.__tracepoints_lock__
+    tracepoints = mod.__tracepoints_specs__
+    funcs = mod.__tracepoints_funcs__
+    spec = lock(mod_lock) do
+        tracepoints[category][kind]
+    end
+    validate_abi(spec, abi, incompatible) || return
+    ptr = generate_probe_fptr(f, spec)
+    lock(mod_lock) do
+        spec.payload[] = ptr
+        funcs[category][kind] = f
+    end
+end
+function validate_abi(spec::TracepointSpec, abi::Symbol, incompatible::Symbol)
+    if !(abi in (:fast, :slow, :all))
+        throw(ArgumentError("Invalid value for `abi`: $abi (must be one of: `:slow`, `:fast`, `:all`)"))
+    end
+    spec_abi = determine_abi(spec)
+    spec_abi_sym = spec_abi == :Any ? :slow : :fast
+    if abi != :all && spec_abi_sym != abi
```

Comment:
In reality the `:fast` ABI is a function of the actual arg type, so this only partially validates right?

Reply:
Yes, this is only intended to validate that the probe supports any fast ABI, the slow ABI, or both as necessary. I'm hoping to avoid specifying which specific fast ABI subsets a probe supports, although I don't know how realistic that is - thoughts?